src/dbus/dbus_message_helper.hpp

```diff
@@ -0,0 +1,741 @@
+
+    if (!aValue.empty())
+    {
+        const int64_t *buf = &aValue.front();
+        VerifyOrExit(dbus_message_iter_append_fixed_array(&subIter, DBUS_TYPE_INT64, &buf,
+                                                          static_cast<int>(aValue.size())) == true);
+    }
+    VerifyOrExit(dbus_message_iter_close_container(aIter, &subIter) == true);
+
+    err = OTBR_ERROR_NONE;
+
+exit:
+    return err;
+}
+
+template <size_t I, typename... Args> struct ElementType
+{
+    using ValueType         = typename std::tuple_element<I, std::tuple<Args...>>::type;
+    using NonconstValueType = typename std::remove_cv<ValueType>::type;
+    using RawValueType      = typename std::remove_reference<NonconstValueType>::type;
+};
+
+template <size_t I, size_t N, typename... Args> class DBusMessageIterFor
+{
+public:
+    static otbrError ConvertToTuple(DBusMessageIter *aIter, std::tuple<Args...> &aValues)
+    {
+        using RawValueType = typename ElementType<N - I, Args...>::RawValueType;
+        RawValueType &val  = std::get<N - I>(aValues);
+        otbrError     err  = DBusMessageExtract(aIter, val);
+
+        SuccessOrExit(err);
+        err = DBusMessageIterFor<I - 1, N, Args...>::ConvertToTuple(aIter, aValues);
+    exit:
+        return err;
+    }
+
+    static otbrError ConvertToDBusMessage(DBusMessageIter *aIter, const std::tuple<Args...> &aValues)
+    {
+        otbrError err = DBusMessageEncode(aIter, std::get<N - I>(aValues));
+
+        SuccessOrExit(err);
+        err = DBusMessageIterFor<I - 1, N, Args...>::ConvertToDBusMessage(aIter, aValues);
+    exit:
+        return err;
+    }
+};
+
+template <size_t N, typename... Args> class DBusMessageIterFor<1, N, Args...>
+{
+public:
+    // XXX: not sure how goto will affect compiler's inline
+    static otbrError ConvertToTuple(DBusMessageIter *aIter, std::tuple<Args...> &aValues)
+    {
+        using RawValueType = typename ElementType<N - 1, Args...>::RawValueType;
+        RawValueType &val  = std::get<N - 1>(aValues);
+        otbrError     err  = DBusMessageExtract(aIter, val);
+
+        return err;
+    }
+
+    static otbrError ConvertToDBusMessage(DBusMessageIter *aIter, const std::tuple<Args...> &aValues)
+    {
+        otbrError err = DBusMessageEncode(aIter, std::get<N - 1>(aValues));
+
+        return err;
+    }
+};
+
+template <typename... Args> otbrError ConvertToDBusMessage(DBusMessageIter *aIter, const std::tuple<Args...> &aValues)
+{
+    return DBusMessageIterFor<sizeof...(Args), sizeof...(Args), Args...>::ConvertToDBusMessage(aIter, aValues);
+}
+
+template <typename... Args>
+constexpr decltype(auto) ConvertToTuple(DBusMessageIter *aIter, std::tuple<Args...> &aValues)
+{
+    return DBusMessageIterFor<sizeof...(Args), sizeof...(Args), Args...>::ConvertToTuple(aIter, aValues);
+}
+
+template <typename... Args> otbrError DBusMessageToTuple(DBusMessage *aMsg, std::tuple<Args...> &aValues)
```

Comment:
Is it possible to change this to `DBusMessage * DBusMessageFromTuple(std::tuple<Args...> &aValues)`?